### DISCUSSION TITLE
Using Fonts

### DISCUSSION THREAD

**Discussion by Marvin1DE** [RudolphRiedel/FT800-FT813#118] [Q&A] [answered]
Hey nochmal, langsam wird das echt viel von meiner Seite aus, sorry erstmal dafür.

Ich bin derzeit dabei die Fonts zu laden für das Project.
Da ich die datein von einer SD Karte lese & bereits Bilder laden konnte habe ich mich jetzt an die Fonts gesetzt,
Die endung muss .bin sein damit ich sie lesen kann (von der sd) in dem fall habe ich sie umbenannt -> a.bin (font.glyph) & b.bin (font.xfont)
also die erste frage
1. ist die endung hier relevant?

Dann habe ich diese in auf den flash geschoben und laut meiner liste müsste sie auch drin sein besetzt nun den platz:
a.bin | 4096 | 137600 -> glyph
b.bin | 141696 | 1912 -> xfont
und andere datein (bilder)..... -> bilder funktionieren

die datein sind nun im flash, also lese ich xfont nochmal zum test in den ram_g auf pos 0 (in der main funktion):
EVE_cmd_flashread(0,141696,1912);
danach setze ich das neue font auf font 30 (in der main funktion):
EVE_cmd_setfont2(30,0,32);   // **benutze große utf-8 range welche ist der erst char? hab 0 und 32 getestet ohne erfolg**

![grafik](https://github.com/RudolphRiedel/FT800-FT813/assets/118083130/9add796b-39b6-44f8-922d-2bd88f103247)

die glyph hab ich zuerst in den flash dann die xfont, nun habe ich ja die neue font gesetzt und nutze diese in tft.c 
![grafik](https://github.com/RudolphRiedel/FT800-FT813/assets/118083130/13b0ec37-b26e-48af-aceb-a28cfce816ef)
aber sobald ich auf die seite mit einem ü wechsel, hängt sich alles auf. (so wie davor ohne die fonts)

mache also immernoch irgendetwas falsch
ich bedanke mich für jegliche hilfe.

edit: benutze Microchip Studio und habe "Auto-detect UTF-8 encoding without signature" aktiv:
![grafik](https://github.com/RudolphRiedel/FT800-FT813/assets/118083130/840704e7-613e-4832-9a52-4507c6ed1081)

**Comment by Marvin1DE** [comment]
habs geschafft das jetzt die .xfont & .glyph datei ohne änderung auf dem flash speicher ist, jedoch geht es immernoch nicht hab mir mal in eba die flash utilities angeschaut und aus spaß eins erstellt -> schreibe ja von sd direkt in den ram_g und dann flash ich es, jedoch gibts dort eine datei namens flash.edf, ist diese relevant für die font? Billder gehen ja wie gesagt und die zeichensätze müssten eigentlich auf dem flash sein

**Comment by RudolphRiedel** [comment]
Moin, mein erster Gedanke war, in der .xfont steht die Basis-Adresse der Glyphen-Daten drin, wenn man die .glyph verschiebt muss man die .xfont anpassen, also so im Speicher nach dem Kopieren aus dem Flash.
Wenn Du das aber so wie gezeigt konvertiert hast und dann Deine a.bin und b.bin so im Flash abgelegt hast, dann passt das.

>danach setze ich das neue font auf font 30 (in der main funktion):

Das ist meine ich der Fehler, CMD_SETFONT2() generiert Display-Listen Kommandos, entsprechend muss das auch im Kontext einer Display-Liste ausgeführt werden.
Normalerweise gebe ich dem Kommando mit das ich ab 32 haben möchte, alles vor dem Leerzeichen ist ohnehin nicht relevant.

Der EVE Asset Builder den Du verwendest sieht übrigens ziemlich alt aus.

>edit: benutze Microchip Studio

Das habe ich zuletzt auch benutzt, wenn das Encoding falsch ist kommen aber in der Regel auch nur die Umlaute kaputt raus.

>jedoch gibts dort eine datei namens flash.edf, ist diese relevant für die font?

Die ist für nichts relevant, das ist eine Hilfsdatei mit deren Hilfe man den Inhalt des Flash parsen können soll, also im Grunde genommen das gleiche wie die .map, nur Maschinen lesbar.

**Reply by Marvin1DE** [reply]
![grafik](https://github.com/RudolphRiedel/FT800-FT813/assets/118083130/3c480857-d87d-41e9-ba38-c4d0f24bcb4c)

oh hab glaube ich vergessen EVE_cmd_setbitmap zu benutzen kurze fragen:
1. als adresse Glyph oder xfont?
2. was als höhe und breite etwa was ich bei eve ausgewählt habe?
3. hab den zeichensatz mit ASTC 12x12 formatiert, muss ich da irgendwas beachten?
in tft.c
![grafik](https://github.com/RudolphRiedel/FT800-FT813/assets/118083130/59f78445-bade-4118-a433-428b3864151a)
in main.c
![grafik](https://github.com/RudolphRiedel/FT800-FT813/assets/118083130/ac36739f-4df7-411c-92ea-681cf605d08b)
Oder sind das 4 möglichkeiten eine neue font zu setzen?

**Reply by RudolphRiedel** [reply]
Mit EVE_cmd_setfont2() brauchst Du kein EVE_cmd_setbitmap() mehr.

CMD_SETBITMAP lässt den co-processor BITMAP_SOURCE, BITMAP_LAYOUT und so weiter in die Display-Liste schreiben.
Und CMD_SETFONT benutzt das dann.
CMD_SETFONT2 enthält sozusagen CMD_SETBITMAP und CMD_SETFONT.

Die .xfont Datei ist eigentlich eine Struktur mit allen notwendigen Daten, siehe Seite 100 vom Programming Guide 2.4.
Interessant ist da drin noch start_of_graphic_data, das kannst Du mal auslesen, sollte aber bereits passen wenn Du die .glyph Datei an 4096 im Flash geschrieben hast so wie Du das im EAB eingestellt hattest.

**Reply by Marvin1DE** [reply]
an sich wird stand jetzt die neue Schriftart benutzt, jedoch ist sobald ich "äüö" oder andere nicht in ascii vorhanden schriftzeichen benutze ein crash bzw der hängt sich auf, und die schriftart ist leicht verschwommen hmm hast du eine Ahnung woran das liegen könnte?

**Reply by Marvin1DE** [reply]
oder doch etwas anders es sieht so aus als wären die zeichen verrutscht wenn ich ein "Œ" mache kommt ein "ū", bin etwas verwirrt wie das zu stande kommt, also die zeichen sind da aber nur falsch angeordnet bzw hab jetzt aus spaß in vs code die tft.c bearbeitet und utf-8 formatiert und dort ein ü eingefügt jetzt gehts hmm liegt wohl an microchip studio 

Edit: Also es ist halt leicht verschwommen und wenn setfont2 benutzt wird gehen die bilder aufeinmal nicht mehr, die auf dem selben display sind hmm...

**Reply by RudolphRrr** [reply, marked as answer]
Also ja, UTF-8 Kodierung ist das eine, das kann man auch für das Studio einstellen, aus dem Kopf weiß ich aber auch gerade nicht genau wo, irgendwas in den Settings mit Editor.
Sonst bekommt man eben irgendeine Windows Codepage.

Und Bilder und Display und so mit Font aus dem Flash, da muss man etwas vorsichtig sein, die Bandbreite vom externen Flash ist eben begrenzt, vor allem wird das bei wahlfreiem Zugriff nicht besser.
Dagegen hilft zum Beispiel den Fontcache zu konfigurieren, der lässt sich aber nur mit einem Font nutzen.
Solange noch Platz im RAM_G ist, vorzugsweise den nutzen.
